src/openms/source/ANALYSIS/MAPMATCHING/PoseClusteringAffineSuperimposer.cpp

```diff
@@ -840,43 +812,49 @@ namespace OpenMS
     setProgress(++actual_progress);
 
     //**************************************************************************
-    // Step 1: Select the most abundant data points only.  After that, disallow modifications
-    //         (we tend to have annoying issues with const_iterator versus
-    //         iterator).
+    // Step 1: Select the most abundant data points only.
     //**************************************************************************
-    PeakPointerArray_ model_map_ini(map_model.begin(), map_model.end());
-    const PeakPointerArray_& model_map(model_map_ini);
-    PeakPointerArray_ scene_map_ini(map_scene.begin(), map_scene.end());
-    const PeakPointerArray_& scene_map(scene_map_ini);
+    // use copy to truncate
+    std::vector<Peak2D> model_map(map_model);
+    std::vector<Peak2D> scene_map(map_scene);
     {
       // truncate the data as necessary
       const Size num_used_points = (Int) param_.getValue("num_used_points");
-      if (model_map_ini.size() > num_used_points)
+
+      // sort the last data points by ascending intensity (from the right, using reverse iterators)
+      //  -> linear in complexity, should be faster than sorting and then taking cutoff
+      //  TODO code-review this
+      if (model_map.size() > num_used_points)
       {
-        model_map_ini.sortByIntensity(true);
-        model_map_ini.resize(num_used_points);
+        std::nth_element(model_map.rbegin(), model_map.rbegin() + (model_map.size() - num_used_points),
+            model_map.rend(), Peak2D::IntensityLess());
+        model_map.resize(num_used_points);
```

Comment:
yes, but we dont *need* sorting. we just want to keep the _n_ most intense elements  and remove the rest (in this case n=num_used_points). As long as we are guaranteed that all elements we want to keep are one one side and all the ones we want to throw away are on the other side of our cutoff, we are fine. Afterwards the array is sorted again by m/z